README.rst
carehare
========

Asyncio RabbitMQ client that handles all the edge cases.

Installation
------------

``pip install carehare``

Usage
-----

Consumer (``async for``)::

    try:
        async with carehare.connect("amqps://guest:guest@localhost") as connection:
            try:
                await connection.queue_declare("my-queue", exclusive=True)
                async with connection.acking_consumer("my-queue") as consumer:
                    async for message in consumer:
                        print(repr(message))
                        # ... if we raise an exception here, we won't ack.
                        #
                        # ... if we `break` from this loop, we won't ack.
                        #
                        # Call `consumer.close()` before (or instead of) `break`
                        # to abort iteration.
            except carehare.ChannelClosedByServer:
                logger.info("RabbitMQ told this one consumer to go away")
            except carehare.ConnectionClosed:
                # Either RabbitMQ is telling us an error (and the outer context
                # manager will throw it), or we called connection.close()
                # ourselves (so we want to close).
                pass
    except carehare.ConnectionClosedByServer:
        # str(error) will give the RabbitMQ error message
        logger.error("RabbitMQ closed our connection")
    except carehare.ConnectionClosedByHeartbeatMonitor:
        logger.error("RabbitMQ went away")

Consumer (``next_delivery``)::

    try:
        async with carehare.connect("amqps://guest:guest@localhost") as connection:
            try:
                await connection.queue_declare("my-queue", exclusive=True)
                async with connection.acking_consumer("my-queue") as consumer:
                    while True:
                        message, delivery_tag = await consumer.next_delivery()
                        # You must ack() (with no await). If RabbitMQ doesn't
                        # receive this ack, it may deliver the same message to
                        # another client.
                        consumer.ack(delivery_tag)
                        if message.startswith(b"okay, go away now"):
                            break
            except carehare.ChannelClosedByServer:
                logger.info("RabbitMQ told this one consumer to go away")
            except carehare.ConnectionClosed:
                # Either RabbitMQ is telling us an error (and the outer context
                # manager will throw it), or we called connection.close()
                # ourselves (so we want to close).
                pass
    except carehare.ConnectionClosedByServer:
        # str(error) will give the RabbitMQ error message
        logger.error("RabbitMQ closed our connection")
    except carehare.ConnectionClosedByHeartbeatMonitor:
        logger.error("RabbitMQ went away")

Publisher::

    try:
        async with carehare.connect("amqps://guest:guest@localhost") as connection:
            try:
                await connection.publish(b"Hello, world!", routing_key="my-queue")
            except carehare.ServerSentNack:
                logger.warn("Failed to publish message")
            except carehare.ChannelClosedByServer:
                # str(err) will give the RabbitMQ error message -- for instance,
                # "404 NOT_FOUND" if the exchange does not exist
                logger.error("Problem with the exchange")
    except carehare.ConnectionClosedByServer:
        # str(error) will give the RabbitMQ error message
        logger.error("RabbitMQ closed our connection")
    except carehare.ConnectionClosedByHeartbeatMonitor:
        logger.error("RabbitMQ went away")

Design decisions
----------------

``carehare`` is designed to turn RabbitMQ's asynchronous error system into
_understandable_ Python exceptions.

Channels
~~~~~~~~

Carehare doesn't let you control RabbitMQ Channels. They aren't Pythonic. (In
RabbitMQ, an exception on a channel closes the channel -- and cancels all its
pending operations.)

Instead, carehare uses channels to handle errors. For instance, Queue.Declare
costs three operations: Channel.Open, Queue.Declare, Channel.Close. Since the
operation has its own channel, it won't interfere with other operations if it
causes an exception.

There's a speed-up for publishing: we lazily open a Channel per *exchange*.
Error codes like "not found", "access refused" and "not implemented" will make
carehare raise an exception on all pending publishes on the same exchange. Don't
worry: a normal "Nack" ("message wasn't delivered") will only make your single
message fail.

Exceptions
~~~~~~~~~~

"Exceptions" are hardly exceptional: as a programmer, they are your job. These
ones are designed to help you solve them.

Connection methods return ``asyncio.Future`` objects. You must await each one
and handle its errors.

Even though you're using Python async context managers, exceptions can't
happen *everywhere*. Carehare will only raise when you ``await`` a response
from RabbitMQ. 

In particular, ``consumer.ack()`` will never raise! You must call it from the
main event loop, but you won't await it.

To code safely, catch these exceptions religiously:

* ``carehare.ChannelClosedByServer``: RabbitMQ did not like the command you
  just ran. Read the exception message for details. After you receive this
  message, you may continue using the RabbitMQ connection.
* ``carehare.ConnectionClosed``: When the connection shuts down, every pending
  ``Future`` will raise this. Only the actual ``Connection`` context manager
  will raise the underlying exception: a ``carehare.ConnectionClosedByServer``
  with a descriptive error message.

Carehare won't raise ``asyncio.Cancelled``.

Back-pressure
~~~~~~~~~~~~~

The core logic is synchronous. It's simpler to reason about. The downside:
neither RabbitMQ nor users will wait for buffers to empty before sending more
data.

Use application-level logic to make sure you don't run out of memory:

* Consuming? Don't worry. Use ``prefetch_count`` to limit the number of messages
  RabbitMQ sends. Always ack: carehare won't permit ``no-ack``.
* Publishing? Carehare forces "publisher confirms", so each publish returns a
  ``Future``. Your application is responsible for not calling publish() too many
  times simultaneously. Use an ``asyncio.Semaphore`` or reason about your
  specific use (for instance, "my server will host max 100 clients, and each
  client can only publish one message at a time").

Comparison to other async RabbitMQ clients
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Other clients tend to try and achieve "RabbitMQ in Python". They fail when it
comes to exceptions:

* ``aiormq``: If your code generates an error during consume, aiormq (4.1.1)
  will catch it and ignore it -- stalling your program.
* ``aioamqp``: If your connection produces an unexpected error, aioamqp will
  catch it and ignore it -- stalling your program. Also, the latest release was
  in 2019.

This author believes it's too confusing to model RabbitMQ's API in Python.
Instead, carehare models your *intent* in Python.

Dependencies
------------

You'll need Python 3.8+ and a RabbitMQ server.

If you have Docker, here's how to start a development server::

    test-server/prepare-certs.sh  # Create SSL certificates used in tests
    docker run --rm -it \
         -p 5671:5671 \
         -p 5672:5672 \
         -p 15672:15672 \
         -v "/$(pwd)"/test-server:/test-server \
         -e RABBITMQ_SSL_CACERTFILE=/test-server/ca.cert \
         -e RABBITMQ_SSL_CERTFILE=/test-server/server.cert \
         -e RABBITMQ_SSL_KEYFILE=/test-server/server.key \
         -e RABBITMQ_SSL_VERIFY=verify_peer \
         -e RABBITMQ_SSL_FAIL_IF_NO_PEER_CERT=true \
         -e RABBITMQ_CONFIG_FILE=/test-server/rabbitmq \
         rabbitmq:3.8.11-management-alpine

During testing, see the RabbitMQ management interface at http://localhost:15672.

Contributing
------------

To add features and fix bugs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First, start a development RabbitMQ server (see above).

Now take on the development cycle:

#. ``tox`` # to ensure tests pass.
#. Write new tests in ``tests/`` and make sure they fail.
#. Write new code in ``carehare/`` to make the tests pass.
#. Submit a pull request.

To deploy
~~~~~~~~~

Use `semver <https://semver.org/>`_.

#. ``git push`` and make sure Travis tests all pass.
#. ``git tag vX.X.X``
#. ``git push --tags``

TravisCI will push to PyPi.


CHANGELOG.md
v1.0.2 - 2022-01-01
~~~~~~~~~~~~~~~~~~~

* Fix mypy error
* Don't log INFO during `Protocol.connection_lost()` [#5]

v1.0.1 - 2021-06-08
~~~~~~~~~~~~~~~~~~~

* Nix errant `print()`.

v1.0.0 - 2021-06-07
~~~~~~~~~~~~~~~~~~~

* Support non-default vhosts. (Before, there was a "/" too many.) [#2]
  This is a major-version bump.
* Warn (and choose default vhost) if users end their amqp:// URLs with "/".

v0.0.11 - 2021-05-14
~~~~~~~~~~~~~~~~~~~~

* `Connection.publish()`: raise `ConnectionClosed` instead of stalling when the
  connection has been closed before `publish()` was called.

v0.0.10 - 2021-03-04
~~~~~~~~~~~~~~~~~~~~

* Nix assertion that fails with uvloop
* Fix send and receive of zero-length message

v0.0.9 - 2021-03-04
~~~~~~~~~~~~~~~~~~~

No changes

v0.0.8 - 2021-02-28
~~~~~~~~~~~~~~~~~~~

* Publish large messages in `frame_max`-sized chunks

v0.0.7 - 2021-02-26
~~~~~~~~~~~~~~~~~~~

* Add debug logging to heartbeat monitor.

v0.0.6 - 2021-02-25
~~~~~~~~~~~~~~~~~~~

* Nix errant `print()`.

v0.0.5 - 2021-02-25
~~~~~~~~~~~~~~~~~~~

* Raise `ValueError` if user asks for SSL "amqp://" or non-SSL "amqps://".
  This conforms with the [URI spec](https://www.rabbitmq.com/uri-spec.html).

v0.0.4 - 2021-02-24
~~~~~~~~~~~~~~~~~~~

* Add `connection.queue_unbind()`

v0.0.3 - 2021-02-21
~~~~~~~~~~~~~~~~~~~

* Fix heartbeat-monitoring logic to avoid disconnect when beats happen at the
  exact wrong times.

v0.0.2 - 2021-02-19
~~~~~~~~~~~~~~~~~~~

* Add `connection.connect()` and `connection.closed`, for users who want to
  skip the async context manager.

v0.0.1 - 2021-02-18
~~~~~~~~~~~~~~~~~~~

* Initial release


carehare/_connection.py
from __future__ import annotations

import asyncio
import ssl
import warnings
from functools import partial
from typing import Optional, Union, cast
from urllib.parse import unquote, urlparse

import pamqp.commands
import pamqp.common

from ._consume_channel import ConsumeChannel
from ._protocol import Protocol


def url_path_to_vhost(path: str) -> str:
    if path == "":
        return "/"  # default vhost
    elif path == "/":
        warnings.warn(
            'Your RabbitMQ URL ends with a slash, which implies vhost "". This is, surprisingly, an error. carehare is interpreting this to mean "default vhost". To avoid this warning, remove the "/" from the end of your URL.',
            DeprecationWarning,
        )
        return "/"  # default vhost
    elif "/" in path[1:]:
        raise ValueError(
            'Invalid AMQP URL path %r: path includes unescaped "/"' % (path,)
        )
    else:
        return unquote(path[1:])


class Connection:
    def __init__(
        self,
        url: str,
        *,
        ssl: Union[ssl.SSLContext, None] = None,
        connect_timeout: Optional[float] = None,
    ):
        self._url = url
        self._ssl = ssl
        self._connect_timeout = connect_timeout

    async def connect(self) -> None:
        """Alternative to `async with` syntax.

        Return if we are connected. Raise if connect fails.

        Do not call this twice on one connection.
        """
        url = urlparse(self._url)
        if url.scheme not in {"amqp", "amqps"}:
            raise ValueError(
                "url must start with 'amqp://' or 'amqps://'; got: %r" % self._url
            )

        if self._ssl is not None and url.scheme == "amqp":
            raise ValueError(
                "When `url` starts with 'amqp://', you must not provide `ssl`. Either omit the `ssl` argument, or make `url` start with 'amqps://'"
            )

        addr = url.netloc.split("@")[-1].split(":")
        if len(addr) == 1:
            host = addr[0]
            port = {"amqp": 5672, "amqps": 5671}[url.scheme]
        else:
            host = addr[0]
            port = int(addr[1])

        virtual_host = url_path_to_vhost(url.path)

        transport, protocol = await asyncio.wait_for(
            asyncio.get_running_loop().create_connection(
                protocol_factory=partial(
                    Protocol,
                    username=url.username,
                    password=url.password,
                    virtual_host=virtual_host,
                ),
                host=host,
                port=port,
                ssl=self._ssl or url.scheme == "amqps",
            ),
            timeout=self._connect_timeout,
        )
        self._protocol = cast(Protocol, protocol)
        await self._protocol.open

    async def __aenter__(self):
        await self.connect()
        return self

    @property
    def closed(self) -> asyncio.Future[None]:
        """Future that resolves when we are closed.

        If closing fails, this Future will have an exception.
        """
        return self._protocol.closed

    async def close(self):
        """Alternative to `async with` syntax.

        You may call this multiple times on a single connection.
        """
        self._protocol.send_close_if_allowed()
        await self.closed

    async def __aexit__(self, exc_type, exc_value, traceback):
        await self.close()  # or raise an exception atop the original exception

    def acking_consumer(
        self, queue_name: str, *, prefetch_size: int = 0, prefetch_count: int = 1
    ) -> ConsumeChannel:
        """Consume messages asynchronously.

        Usage:

            async with connection.acking_consumer("my-queue") as consumer:
                async for message in consumer:
                    print("hi! %r" % message)

        After the inner block completes, the consumer will ack the message.

        If your inner block raises an exception, the message will _not_ be
        acked. Once the outer block exits, RabbitMQ may deliver the message
        to another consumer.

        You may call `consumer.close()` to stop the flow of messages in the
        `async for` loop. You probably shouldn't call it from _within_ the
        `async for` loop because a low-traffic queue won't yield to your code
        in a timely manner.

        `prefetch_size` and `prefetch_count` hint to RabbitMQ how much of a
        backlog should be on this queue. Smaller numbers are "fairer" because a
        slow consumer won't hold on to many messages. Larger numbers are
        "faster" because messages will be available locally before your acks
        reach the server.
        """
        return self._protocol.acking_consumer(
            queue_name=queue_name,
            prefetch_size=prefetch_size,
            prefetch_count=prefetch_count,
        )

    def publish(
        self, message: bytes, *, exchange_name: str = "", routing_key: str = ""
    ) -> asyncio.Future[None]:
        return self._protocol.publish(
            message, exchange_name=exchange_name, routing_key=routing_key
        )

    def exchange_declare(
        self,
        exchange_name: str,
        *,
        exchange_type: str = "direct",
        passive: bool = False,
        durable: bool = False,
        auto_delete: bool = False,
        internal: bool = False,
        arguments: pamqp.common.Arguments = None,
    ) -> asyncio.Future[None]:
        return self._protocol.rpc(
            pamqp.commands.Exchange.Declare(
                exchange=exchange_name,
                exchange_type=exchange_type,
                passive=passive,
                durable=durable,
                auto_delete=auto_delete,
                internal=internal,
                nowait=True,
                arguments=arguments,
            )
        )

    def queue_bind(
        self,
        queue_name: str,
        exchange_name: str,
        *,
        routing_key: str = "",
        arguments: pamqp.common.Arguments = None,
    ) -> asyncio.Future[None]:
        return self._protocol.rpc(
            pamqp.commands.Queue.Bind(
                queue=queue_name,
                exchange=exchange_name,
                routing_key=routing_key,
                nowait=True,
                arguments=arguments,
            )
        )

    def queue_declare(
        self,
        queue_name: str,
        *,
        passive: bool = False,
        durable: bool = False,
        exclusive: bool = False,
        auto_delete: bool = False,
        arguments: pamqp.common.Arguments = None,
    ) -> asyncio.Future[None]:
        return self._protocol.rpc(
            pamqp.commands.Queue.Declare(
                queue=queue_name,
                passive=passive,
                durable=durable,
                exclusive=exclusive,
                auto_delete=auto_delete,
                nowait=True,
                arguments=arguments,
            )
        )

    def queue_unbind(
        self,
        queue_name: str,
        exchange_name: str,
        *,
        routing_key: str = "",
        arguments: pamqp.common.Arguments = None,
    ) -> asyncio.Future[None]:
        return self._protocol.rpc(
            pamqp.commands.Queue.Unbind(
                queue=queue_name,
                exchange=exchange_name,
                routing_key=routing_key,
                arguments=arguments,
            ),
            ignored_frame_types=frozenset([pamqp.commands.Queue.UnbindOk]),
        )


def connect(
    url: str,
    *,
    ssl: Union[ssl.SSLContext, None] = None,
    connect_timeout: Optional[float] = None,
) -> Connection:
    return Connection(url, ssl=ssl, connect_timeout=connect_timeout)
